[WebDriver] Add confirmation of socket connection before sending commands to the backend

haruhisa-shin opened the pull request:
#### 827066c642dd0139614ce43da695a13630da596f
<pre>
[WebDriver] Add confirmation of socket connection before sending commands to the backend
<a href="https://bugs.webkit.org/show_bug.cgi?id=274509">https://bugs.webkit.org/show_bug.cgi?id=274509</a>

Reviewed by Carlos Garcia Campos.

The SessionHost of WebDriver is trying to send a command to backend
without checking whether the socket is connected or not.
If the socket is disconnected, the completionHandler will remain added
to the queue and will not be called by anyone, and the process will not
be completed.
To solve this problem, add code to check the status of the socket
before sending the command.

* Source/WebDriver/SessionHost.cpp:
(WebDriver::SessionHost::sendCommandToBackend):

Canonical link: <a href="https://commits.webkit.org/279417@main">https://commits.webkit.org/279417@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d905d113dffaea3b824c3ad2fc7818ace37f4bdb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53339 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32690 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5838 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56619 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4065 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/40218 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3832 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43245 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2662 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55437 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/30970 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46072 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24376 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/27833 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3392 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2221 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49595 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3561 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58214 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28489 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/3568 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/50643 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29703 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46280 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/49978 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11646 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30627 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29467 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebDriver/SessionHost.cpp:
```diff
@@ -41,6 +41,11 @@ void SessionHost::inspectorDisconnected()
 
 long SessionHost::sendCommandToBackend(const String& command, RefPtr<JSON::Object>&& parameters, Function<void (CommandResponse&&)>&& responseHandler)
 {
+    if (!isConnected()) {
```
So, this is not specific to the socket implementation, can you update the bug title and commit message, pelase?

haruhisa-shin replied:
@carlosgcampos I updated them. Thank you for your review!

haruhisa-shin committed:
[WebDriver] Add confirmation of socket connection before sending commands to the backend
https://bugs.webkit.org/show_bug.cgi?id=274509

Reviewed by Carlos Garcia Campos.

The SessionHost of WebDriver is trying to send a command to backend
without checking whether the socket is connected or not.
If the socket is disconnected, the completionHandler will remain added
to the queue and will not be called by anyone, and the process will not
be completed.
To solve this problem, add code to check the status of the socket
before sending the command.

* Source/WebDriver/SessionHost.cpp:
(WebDriver::SessionHost::sendCommandToBackend):

Canonical link: https://commits.webkit.org/279417@main
